### PR TITLE
Move clientconfig into clientv3 so that it can be reused by both etcd…

### PR DESCRIPTION
**Pull request by ahrtr**
Part of [13624](https://github.com/etcd-io/etcd/issues/13624), and it's one of the follow-ups to PR [13635](https://github.com/etcd-io/etcd/pull/13635).

Move the original [clientConfig](https://github.com/etcd-io/etcd/blob/main/etcdctl/ctlv3/command/global.go#L100) from etcdctl to clientv3, so that it can be reused by both etcdctl and v3 discovery.

Note that etcd uses golang's standard package `flag` to process command line flags, while etcdctl uses` github.com/spf13/pflag`, and etcdctl and v3 discovery has different flag names, such as `--endpoints` vs `--discovery-endpoints`. So it's difficult to share the same code to process flags between them. So the command line flags keep unchanged in this PR.

cc @serathius @ptabor @spzala

### CODE REVIEW

**Comment by ahrtr** [comment]
All test are green now.

File: client/v3/config.go [hunk 1]
```diff
@@ -90,3 +90,28 @@ type Config struct {
 
 	// TODO: support custom balancer picker
 }
+
+type ClientConfig struct {
+	Endpoints        []string      `json:"endpoints"`
+	RequestTimeOut   time.Duration `json:"request-timeout"`
```
Comment (serathius) [inline client/v3/config.go:96]:
```suggestion
	RequestTimeout   time.Duration `json:"request-timeout"`
```

Comment (ahrtr) [reply client/v3/config.go]: done.

**Comment by serathius** [comment]
Overall looks much better than previous PR, with clearer PR we can see one additional problem clarity of public client interface. `client/v3` package already includes `Config` struct for configuration of client library, having additional `ClientConfig` inside makes things confusing. 

As a really good refactor I would normally merge it as a gradual improvement, however with this adding to public package interface I'm not sure if we can leave things as they are without long term plan how client configuration should look at in the end. This is because I expect that this package cannot introduce a breaking change in configuration. `client.Config` and `client.ClientConfig` are similar enough that are sure to confuse users.

I see two things:
* Propose a new version of `client.Config` that can be used for discovery.
* Postpone the merge and find a less user facing package for `ClientConfig`

cc @spzala @ptabor

**Comment by ahrtr** [comment]
Yes, it's a valid concern.  We have two solutions for now,
1. We do not need this PR, keep everything it as it's, which means that etcdctl and v3 discovery will have similar/duplicate clientConfig. 
2. Add comment to explain the difference between Config and ClientConfig. Currently ClientConfig is only used by etcdctl and v3 discovery. If users want to develop their own applications which need to process command-line flags, then ClientConfig may be useful as well. We could also move the ClientConfig to a separate file such as client_config.go. If we follow this approach, we can't clarity refactor on the Config (such as merge some common fields between Config and ClientConfig), because it may have already been used by some users' applications. 

Please let me know if you have any better proposal.

**Comment by ptabor** [comment]
The crucial different between: 

  "Config" and "ClientConfig" is that "Config" is mostly a resolved configuration, i.e. it contains objects likes Loggers, TLS that are not serialisable any longer (and especially directly) and that can be potentially shared between objects. 

The ClientConfig is supposed to be fully declarative configuration that is serializable to JSON or set of flags, and after serialization and deserialization the same exact object can be reached. 

I'm leaning towards: 
 1.  Naming `ClientConfig` as e.g. 'ClientSpec` or just `clientv3.Spec`- such that it's clean that the object is just declarative description. 
 2.  Providing resolution that converts `Config` -> `ClientSpec`, i.e. `NewConfigFromSpec(spec)`

**Comment by ahrtr** [comment]
Thanks @ptabor. 

It makes sense to change `ClientConfig` to `ClientSpec`, it can also mitigate the confusion to users pointed out by @serathius .

But I do not see any need or use case to convert `Config` ->  `ClientSpec` for now. Instead, we should provide resolution that convert `ClientSpec` to `Config`.  It's a very natural process. The command-line flags go to `ClientSpec` firstly, then convert it to `Config` afterwards. Since it may also impact the existing `etcdctl`, so I'd like to do it in a separate PR.  WDYT? @ptabor  @serathius @spzala

**Comment by ahrtr** [comment]
Does it make more sense to rename `ClientConfig` to `ConfigSpec`?  `Spec` is just a generic spec, but `ConfigSpec` is a spec specific to client config.

**Review by serathius — APPROVED** [review]
Great change! Thanks for the followups!

**Comment by ahrtr** [comment]
cc @spzala @ptabor

File: server/etcdserver/api/v3discovery/discovery.go [hunk 2]
```diff
@@ -55,22 +55,8 @@ var (
...
-
-	InsecureTransport  bool   `json:"discovery-insecure-transport"`
-	InsecureSkipVerify bool   `json:"discovery-insecure-skip-tls-verify"`
-	CertFile           string `json:"discovery-cert"`
-	KeyFile            string `json:"discovery-key"`
-	TrustedCAFile      string `json:"discovery-cacert"`
-
-	User     string `json:"discovery-user"`
-	Password string `json:"discovery-password"`
+	clientv3.ConfigSpec `json:"client-spec"`
```
Comment (ptabor) [inline server/etcdserver/api/v3discovery/discovery.go:58]: nit: I would just call it `client` to not confuse whether its `spec` or `config`... but I don't have strong opinion here.

**Review by ptabor — APPROVED** [review]
Thank you.

File: server/etcdserver/api/v3discovery/discovery.go (hunk 2 above)
Comment (ahrtr) [reply server/etcdserver/api/v3discovery/discovery.go]: Updated & thanks.

**Comment by ahrtr** [comment]
All test passed (we can ignore the semaphoreci failure, which is due to dockerhub rate limit). PTAL cc @serathius @ptabor @spzala

**Comment by ahrtr** [comment]
All test passes after re-triggering the pipeline.

**Pull request merged**